src/main/java/org/skriptlang/skript/lang/condition/ConditionalBuilder.java

```diff
@@ -0,0 +1,173 @@
+package org.skriptlang.skript.lang.condition;
+
+import org.jetbrains.annotations.Contract;
+import org.skriptlang.skript.lang.condition.Conditional.Operator;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static org.skriptlang.skript.lang.condition.Conditional.negate;
+
+/**
+ * Builds a DNF {@link CompoundConditional}, meaning it is solely composed of groups of ANDs all ORed together,
+ * ex. {@code (a && !b && c) || b || (!c && d)}.
+ * <br>
+ * A builder should no longer be used after calling {@link #build()}.
+ * @param <T> The context class to use for evaluation.
+ * @see CompoundConditional
+ */
+public class ConditionalBuilder<T> {
+
+	private CompoundConditional<T> root;
+
+	/**
+	 * Creates an empty builder.
+	 */
+	ConditionalBuilder() {
+		this.root = null;
+	}
+
+	/**
+	 * Creates a builder with a single conditional.
+	 * @param root If given a {@link CompoundConditional}, it is used as the root conditional.
+	 *             Otherwise, a OR compound conditional is created as the root with the given conditional within.
+	 */
+	ConditionalBuilder(Conditional<T> root) {
+		if (root instanceof CompoundConditional<T> compoundConditional) {
+			this.root = compoundConditional;
+		} else {
+			this.root = new CompoundConditional<>(Operator.OR, root);
+		}
+	}
+
+	/**
+	 * @return The root conditional, which will be DNF-compliant
+	 * @throws IllegalStateException if the builder is empty.
+	 */
+	public Conditional<T> build() {
+		if (root == null)
+			throw new IllegalStateException("Cannot build an empty conditional!");
+		return root;
+	}
+
+	/**
+	 * Adds conditionals to the root node via the AND operator: {@code (existing) && newA && newB && ...}.
+	 * If the root is currently OR, the statement is transformed as follows to maintain DNF:
+	 * {@code (a || b) && c -> (a && c) || (b && c)}
+	 * @param andConditionals conditionals to AND to the existing conditional.
+	 * @return the builder
+	 */
+	@SafeVarargs
+	@Contract("_ -> this")
+	public final ConditionalBuilder<T> and(Conditional<T>... andConditionals) {
+		if (root == null) {
+			root = new CompoundConditional<>(Operator.AND, andConditionals);
+			return this;
+		}
+
+		// unroll conditionals if they're ANDs
+		List<Conditional<T>> newConditionals = unroll(List.of(andConditionals), Operator.AND);
+
+		// if the root is still just AND, we can just append.
+		if (root.getOperator() == Operator.AND) {
+			root.addConditionals(newConditionals);
+			return this;
+		}
+		// Otherwise, we need to transform:
+		// (a || b) && c -> (a && c) || (b && c)
```

Comment:
Is there a reason to transform these?
e.g. if we have `a OR b` and we are AND-ing with `c`,  why *can't* we just nest the existing root within an AND (as the comment descibes: `(a || b) && c`

Reply:
We could, but it would no longer be DNF and loses some of the nice properties of DNF being a max depth of 2 and always ors of ands.